File: app/storage.py
import os, httpx
from urllib.parse import quote

class Storage:
    def __init__(self, base: str | None = None):
        self.base = base or os.environ.get("INNO_STORAGE_BASE", "http://storage.internal")

    async def _post(self, path: str, body: dict):
        async with httpx.AsyncClient() as c:
            r = await c.post(f"{self.base}{path}", json=body)
            r.raise_for_status()
            return r.json()

    async def query(self, sql: str, params: list | None = None):
        return (await self._post("/_storage/sql/query", {"sql": sql, "params": params or []}))["results"]

    async def execute(self, sql: str, params: list | None = None):
        return await self._post("/_storage/sql/execute", {"sql": sql, "params": params or []})

    async def put_file(self, key: str, data: bytes):
        async with httpx.AsyncClient() as c:
            r = await c.put(f"{self.base}/_storage/files/{quote(key, safe='')}", content=data); r.raise_for_status(); return r.json()

    async def get_file(self, key: str) -> bytes | None:
        async with httpx.AsyncClient() as c:
            r = await c.get(f"{self.base}/_storage/files/{quote(key, safe='')}")
            if r.status_code == 404: return None
            r.raise_for_status(); return r.content

    async def list_files(self) -> list:
        async with httpx.AsyncClient() as c:
            r = await c.get(f"{self.base}/_storage/files"); r.raise_for_status(); return r.json()["keys"]

    async def delete_file(self, key: str):
        async with httpx.AsyncClient() as c:
            r = await c.delete(f"{self.base}/_storage/files/{quote(key, safe='')}"); r.raise_for_status(); return r.json()

def current_user(request) -> dict:
    h = request.headers
    groups = [g.strip() for g in h.get("x-forwarded-groups", "").split(",") if g.strip()]
    return {"email": h.get("x-forwarded-user", ""), "groups": groups}

class NotConnected(Exception):
    def __init__(self, connect_url: str):
        super().__init__(f"not connected — open {connect_url} to link your account")
        self.connect_url = connect_url

# Per-user backend credentials (APP-CONTRACT §2.2). `caller_assertion` is the
# value of THIS request's inbound X-Caller-Assertion header — the app must
# echo it (identity rides only that platform-signed token). Returns a live
# credential ({"access_token": ...} or {"header": ...}), or raises
# NotConnected(connect_url) which the app relays to the user.
class Connections:
    def __init__(self, base: str | None = None, client: httpx.AsyncClient | None = None):
        self.base = base or os.environ.get("INNO_STORAGE_BASE", "http://storage.internal")
        self.client = client

    async def get(self, name: str, caller_assertion: str | None) -> dict:
        headers = {"X-Caller-Assertion": caller_assertion or ""}
        if self.client is not None:
            r = await self.client.post(f"{self.base}/_connections/{name}", headers=headers)
        else:
            async with httpx.AsyncClient() as c:
                r = await c.post(f"{self.base}/_connections/{name}", headers=headers)
        r.raise_for_status()
        out = r.json()
        if out.get("status") == "not_connected":
            raise NotConnected(out.get("connect_url"))
        return {"access_token": out.get("access_token"), "header": out.get("header"), "expires_at": out.get("expires_at")}

File: lib/storage.js
export class Storage {
  constructor(opts = {}) {
    this.base = opts.base || process.env.INNO_STORAGE_BASE || "http://storage.internal";
    this.fetcher = opts.fetcher || fetch;
  }
  async _post(path, body) {
    const r = await this.fetcher(`${this.base}${path}`, { method: "POST", headers: { "Content-Type": "application/json" }, body: JSON.stringify(body) });
    if (!r.ok) throw new Error(`storage_error:${r.status}`);
    return r.json();
  }
  async query(sql, params = []) { return (await this._post("/_storage/sql/query", { sql, params })).results; }
  async execute(sql, params = []) { return this._post("/_storage/sql/execute", { sql, params }); }
  // Cross-app data links (APP-CONTRACT §2.1): the linked source app's D1,
  // available only after the owner created the link and the app redeployed.
  async queryLinked(sourceApp, sql, params = []) { return (await this._post(`/_storage/linked/${encodeURIComponent(sourceApp)}/sql/query`, { sql, params })).results; }
  async executeLinked(sourceApp, sql, params = []) { return this._post(`/_storage/linked/${encodeURIComponent(sourceApp)}/sql/execute`, { sql, params }); }
  async putFile(key, bytes) { const r = await this.fetcher(`${this.base}/_storage/files/${encodeURIComponent(key)}`, { method: "PUT", body: bytes }); if (!r.ok) throw new Error(`storage_error:${r.status}`); return r.json(); }
  async getFile(key) { const r = await this.fetcher(`${this.base}/_storage/files/${encodeURIComponent(key)}`); if (r.status === 404) return null; if (!r.ok) throw new Error(`storage_error:${r.status}`); return new Uint8Array(await r.arrayBuffer()); }
  async listFiles() { const r = await this.fetcher(`${this.base}/_storage/files`); if (!r.ok) throw new Error(`storage_error:${r.status}`); return (await r.json()).keys; }
  async deleteFile(key) { const r = await this.fetcher(`${this.base}/_storage/files/${encodeURIComponent(key)}`, { method: "DELETE" }); if (!r.ok) throw new Error(`storage_error:${r.status}`); return r.json(); }
}
export function currentUser(headers) {
  const get = (k) => (typeof headers.get === "function" ? headers.get(k) : headers[k]) || "";
  return { email: get("x-forwarded-user"), groups: get("x-forwarded-groups").split(",").map((s) => s.trim()).filter(Boolean) };
}

export class NotConnected extends Error {
  constructor(connectUrl) {
    super(`not connected — open ${connectUrl} to link your account`);
    this.name = "NotConnected";
    this.connectUrl = connectUrl;
  }
}

// Per-user backend credentials (APP-CONTRACT §2.2). `callerAssertion` is the
// value of THIS request's inbound `X-Caller-Assertion` header — the app must
// echo it (identity rides only that platform-signed token). Returns a live
// credential ({ access_token } or { header }), or throws NotConnected(connectUrl)
// which the app relays to the user.
export class Connections {
  constructor(opts = {}) {
    this.base = opts.base || process.env.INNO_STORAGE_BASE || "http://storage.internal";
    this.fetcher = opts.fetcher || fetch;
  }
  async get(name, callerAssertion) {
    const r = await this.fetcher(`${this.base}/_connections/${name}`, {
      method: "POST",
      headers: { "X-Caller-Assertion": callerAssertion || "" },
    });
    if (!r.ok) throw new Error(`connections_error:${r.status}`);
    const out = await r.json();
    if (out.status === "not_connected") throw new NotConnected(out.connect_url);
    return { access_token: out.access_token, header: out.header, expires_at: out.expires_at };
  }
}
